src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs

```diff
@@ -310,5 +310,66 @@ public async Task TreeNotReusedIfParseOptionsChangeChangeBetweenRuns()
 
             Assert.NotSame(generatedTreeBeforeChange, generatedTreeAfterChange);
         }
+
+        [Fact]
+        public async Task ChangeToDocumentThatDoesNotImpactGeneratedDocumentReusesDeclarationTree()
+        {
+            using var workspace = CreateWorkspace();
+            var analyzerReference = new TestGeneratorReference(new SingleFileTestGenerator("// StaticContent"));
+            var project = AddEmptyProject(workspace.CurrentSolution)
+                .AddAnalyzerReference(analyzerReference)
+                .AddDocument("RegularDocument.cs", "// Source File", filePath: "RegularDocument.cs").Project;
+
+            // Ensure we already have a compilation created
+            _ = await project.GetCompilationAsync();
+
+            project = await MakeChangesToDocument(project);
+
+            var compilationAfterFirstChange = await project.GetRequiredCompilationAsync(CancellationToken.None);
+
+            project = await MakeChangesToDocument(project);
+
+            var compilationAfterSecondChange = await project.GetRequiredCompilationAsync(CancellationToken.None);
+
+            // When we produced compilationAfterSecondChange, what we would ideally like is that compilation was produced by taking
+            // compilationAfterFirstChange and simply updating the syntax tree that changed, since the generated documents didn't change.
+            // That allows the compiler to reuse the same declaration tree for the generated file. This is hard to observe directly, but if we reflect
+            // into the Compilation we can see if the declaration tree is untouched. We won't look at the original compilation, since
+            // that original one was produced by adding the generated file as the final step, so it's cache won't be reusable, since the
+            // compiler separates the "most recently changed tree" in the declaration table for efficiency.
+
+            var cachedStateAfterFirstChange = GetDeclarationManagerCachedStateForUnchangingTrees(compilationAfterFirstChange);
+            var cachedStateAfterSecondChange = GetDeclarationManagerCachedStateForUnchangingTrees(compilationAfterSecondChange);
+
+            Assert.Same(cachedStateAfterFirstChange, cachedStateAfterSecondChange);
+
+            static object GetDeclarationManagerCachedStateForUnchangingTrees(Compilation compilation)
+            {
+                var syntaxAndDeclarationsManager = compilation.GetFieldValue("_syntaxAndDeclarations");
```

Comment:
💡 IVT to test projects should be acceptable, and we can add a test accessor within the compiler layer to access a private field without reflection.

Reply:
My concern was that simply opened up the _entire_ compiler layer, and we've had mistakes in the past if I recall where we had compiler code being tested entirely at the IDE layer, or could have features accidentally testing compiler internals rather than shipping behavior. @jaredpar any preference here? The backstory:

1. The compiler has optimizations where it optimizes for certain IDE patterns around replacing syntax trees. There's no public API to actually ask whether the optimization happened or not.
2. We want some tests in the IDE which asserts that when we're processing compilations we're keeping those optimizations intact.

Short of building a "Mock" compilation to test the IDE code with, it seems our options are either:

1. Have the test reflect on the bit of the compiler it cares about. In the case if the compiler were to ever change the implementation, the test would just break.
2. Grant IVT to test projects in the IDE.

I'm going with 1, since the hack is targeted, should only break if somebody is tinkering in this area, and should be straightforward to keep the test running.

Reply:
@jaredpar So in this case the question (is this reusing a decl tree?) is _really_ an implementation detail I could imagine the compiler might change. We might be able to do an API where the decl tree creation is more explicit -- the compiler has some opaque API to create it for a tree and then we are able to reuse that.

Reply:
I think if we look around we can find analogous APIs to use for inspiration. Consider performance counter APIs are often giving real info of implementation details like this but it's also exposed publicly.